packages/shared/src/utils/deprecate.ts

```diff
@@ -0,0 +1,91 @@
+type AnyFn = (...args: any[]) => any;
+type AnyClass = new (...args: any[]) => any;
+
+interface DeprecateType {
+  (fnOrPropName: string, warning: string, fnOrMethodOrCls: AnyClass, isStatic?: boolean): void;
+  (fnOrPropName: string, warning: string, fnOrMethodOrCls: AnyFn, isStatic?: boolean): AnyFn;
+}
+
+/**
+ * Mark class properties/methods or functions as deprecated.
+ *
+ * On marked function/methods/class props a console.warn
+ * will be shown ONLY for non-production environments on the
+ * first time the deprecated object will be accessed or invoked.
+ *
+ * Examples
+ * 1. Deprecate class method
+ * class Example {
+ *   getSomething = deprecate('getSomething', 'Use `getSomethingElse` instead.', (arg1, arg2) => {
+ *       return `getSomethingValue:${arg1 || '-'}:${arg2 || '-'}`;
+ *     });
+ *   }
+ *
+ * 2. Deprecate class property
+ * class Example {
+ *   something: string;
+ *   constructor(something: string) {
+ *     this.something = something;
+ *   }
+ * }
+ *
+ * deprecate('something', 'Use `somethingElse` instead.', Example);
+ *
+ * 3. Deprecate class static property
+ * class Example {
+ *   static something: string;
+ * }
+ *
+ * deprecate('something', 'Use `somethingElse` instead.', Example, true);
+ *
+ * 4. Deprecate function
+ * const getSomething = deprecate('getSomething', 'Use `getSomethingElse` instead.', () => {
+ *   return 'getSomethingValue';
+ * });
+ */
+export const deprecate: DeprecateType = (fnOrPropName, warning, fnOrMethodOrCls, isStatic = false) => {
```

Comment:
❓ Can we combine `fnOrPropName`, `fnOrMethodOrCls`, and `isStatic` into a single argument? Classes are functions and they have prototypes. I'd suggest trying to do it using the typeof operator.

Reply:
I am open to suggestions.
The reasons behind this cumbersome method signature:
- `fnOrPropName` and `fnOrMethodOrCls` are required for class properties since you cannot wrap them somehow and the only way I found to decorate them was, by rewriting the property on the class and output a console log when the property accessor is executed.
- `isStatic` is required to be set when a property is static because there is no way to check its existence if there is no default value set.
- if `fnOrPropName` is not passed when decorating the method, the warning should contain the deprecated method name since using the arrow function as `fnOrMethodOrCls` would result in showing a warning with missing information about what should be removed. We could enforce the use of named functions instead of arrow functions but this would cause an issue with decorated functions using `this` inside.

I could try to implement something like this, if it makes more sense (or split the it into 2 wrappers):
```typescript
interface DeprecateType {
    (warning: string, fnOrMethodOrCls: AnyClass, fnOrPropName: string): void;
    (warning: string, fnOrMethodOrCls: AnyFn, isStatic?: boolean): AnyFn;
  }

// method
class Example {
   getSomething = deprecate('Use `getSomethingElse` instead.', (arg1, arg2) => {
       return `getSomethingValue:${arg1 || '-'}:${arg2 || '-'}`;
    });
}

// function
const getSomething = deprecate('Use `getSomethingElse` instead.', (arg1, arg2) => {
    return `getSomethingValue:${arg1 || '-'}:${arg2 || '-'}`;
});

// property
class Example {
    something: string;
}

deprecate('Use `somethingElse` instead.', Example, 'something');


// static property
class Example {
    static something: string;
}

deprecate('Use `somethingElse` instead.', Example, 'something', true);
```
Does anyone have an example of how we could decorate (without using TS decorators) a class property without using fnOrPropName` and `fnOrMethodOrCls`?